### DISCUSSION TITLE
Book matching (experimental)

### DISCUSSION THREAD

**Discussion by advplyr** [advplyr/audiobookshelf#157] [Announcements]
## Re-introducing Book matching

Included in `v1.5.6`
Toggle between Google Books and Open Library to use for searching.
Select a search result and choose which fields you would like to use when updating.

![image](https://user-images.githubusercontent.com/67830747/139325081-7c96414a-e725-43ef-b1ae-d737f723ce37.png)

![image](https://user-images.githubusercontent.com/67830747/139325476-035f705f-e1da-4e24-b251-95e771b4711a.png)

### Things still left to do:
- Allow searching by ISBN
- Show genres found in a multi-select box
- Scanner option to automatically lookup and fill in book details that are empty
- Google books thumbnail covers are very small, requesting the volume info specifically via `https://www.googleapis.com/books/v1/volumes/<id>` gives access to larger covers.

**Comment by jarrodCoombes** [comment]
This looks freaking awesome!!! Can't wait to play with it later today.

**Reply by Jdiesel87** [reply]
Great work! I was going to suggest field selection dialogue during import like Calibre allows for but didn't want to overwhelmed you with requests. Seems like your are already a few steps ahead.

**Comment by jarrodCoombes** [comment]
This works well, generally the fist match is the correct one. But if you do build it into the scanner, I'd like to have the options on which fields it auto populates. For my use I'd prefer the folder be used for author, title and series, but the book march for the cover, the description, year etc.

**Reply by advplyr** [reply]
The folder names and info taken from the filenames will always take priority. I would like to eventually have the order of precedence be adjustable, but I think at first it will need to be preset. Filling in data from external sources will be something you need to enable in settings.

**Comment by Jdiesel87** [comment]
This has been working well for me. I added about a dozen books today and was able to get them all tagged and populated to my liking within a minute or too. This really speeds up the process while still allowing one to be picky about what metadata is getting pulled in.

**Comment by TheSylus** [comment]
Could you also please add the support for Audible.com

In another post someone already referred to this guide, as it is often used in combination with plex.
https://github.com/seanap/Plex-Audiobook-Guide.

In the repo you can directly find the web scrapper (mp3tag/data/sources)
[Plex-Audiobook-Guide-master (1).zip](https://github.com/advplyr/audiobookshelf/files/7446225/Plex-Audiobook-Guide-master.1.zip)

Maybe you can reuse part of the code.

**Reply by advplyr** [reply]
I hesitate to add web scraping to this project because it is unstable. The website could start rate limiting requests then it looks like the feature isn't working. Or they change the DOM on the web page being scraped and it stops working altogether.

If a web scraper is used to get metadata then I would prefer to have it as an external service that audiobookshelf makes requests to, rather then it be in audiobookshelf.

I would like to eventually have support for Audible metadata, I just haven't seen a clean way to do it yet.

**Reply by seanap** [reply]
Don't use the web scraping, it's good enough for tagging since we can modify any errors on the spot.  Check out the Audnexus project which uses the Audible API.  This was written to support a Plex agent but should be able to be integrated.  It's really good, it also scrapes author data.

https://github.com/laxamentumtech/audnexus

**Reply by Jdiesel87** [reply]
Very interesting project. Thanks for posting.

**Reply by advplyr** [reply]
I checked out that project a bit and I didn't see a search api endpoint. It looks like you need to know the audible ASIN to fetch metadata. Am I missing something?

**Reply by seanap** [reply]
It's probably best answered by @djdembeck but for the Plex agent I know there is an api and the input is usually author+title. Here is an example of an author search: https://api.audnex.us/authors?name=AJ%20Banner

**Reply by djdembeck** [reply]
Until I can integrate a proper search backend into the API, author search is the only one currently available. The Plex agent uses Audible's own search API for finding the ASIN: https://github.com/djdembeck/Audnexus.bundle/blob/main/Contents/Code/search_tools.py#L13

**Reply by advplyr** [reply]
Are there any docs for this api?
https://api.audible.com/1.0/catalog/products?response_groups=contributors,product_desc,product_extended_attrs,product_attrs,media,rating&image_sizes=500,1024&keywords=enders%20game&num_results=25&products_sort_by=Relevance

The search with keywords works well and the images are nice quality. The summary is HTML though, and I don't see a way to pull genres from this. It is fast too. Rate limiting can be worked around but I can't find any docs.

**Reply by advplyr** [reply]
I scrolled right past some info on it: https://audible.readthedocs.io/en/latest/misc/external_api.html
I think this api can be integrated. The primary benefit will be nice images.

**Reply by advplyr** [reply]
Audible is supported as a match provider: https://github.com/advplyr/audiobookshelf/discussions/157#discussioncomment-1678587

**Comment by GrapeApe230** [comment]
To piggy back on this a bit. Would it make sense to give users the ability to use their own google books api key? This way we avoid rate limiting or hit count. Maybe provide a link to on how to create your own key next to the option? https://developers.google.com/books/docs/v1/getting_started

**Reply by advplyr** [reply]
Yeah I thought about this, but I didn't see any info on how google rate limits their free requests. Their free search api turned out to work really well as far as accuracy and returning all the details we would need.

**Reply by Jdiesel87** [reply]
Yeah I haven't ran into any issues so far and was hitting it pretty hard at one point

**Reply by Jdiesel87** [reply]
Honestly the only thing really missing now is a good source for genres. Google Books does a good enough job with just about everything minus genres.

**Reply by seanap** [reply]
The Audible API gives the best variety of genres I've come across. They are actually pretty useful. Although Audible isn't the best choice for non-audiobook titles. Sample below from "Jay Allan - Invasion", it provided 5 different Genres:
![image](https://user-images.githubusercontent.com/17012946/141134596-ed96e28c-abf2-4432-be16-d9240eb5040f.png)

**Reply by advplyr** [reply]
Genres are getting scraped from the audible page: https://github.com/laxamentumtech/audnexus/blob/main/src/helpers/books/audibleScrape.ts#L26

**Reply by GrapeApe230** [reply]
Looks like the rate is 1000 per day and 100 queries per minute per user
![image](https://user-images.githubusercontent.com/39072577/141140831-6e9ace65-4848-4b3e-ac95-e2bc086634d5.png)

**Reply by advplyr** [reply]
Can you link to that? I think that is for the endpoints that require an API key

**Reply by GrapeApe230** [reply]
I was grabbing that from my IAM & Admin panel
https://console.cloud.google.com/iam-admin/quotas?service=books.googleapis.com
![image](https://user-images.githubusercontent.com/39072577/141143383-e0b8c147-48b3-4dd1-8ef2-004b2bfa7ab2.png)

**Reply by advplyr** [reply]
Yeah they may hold to that for IP addresses too, but I didn't see that documented anywhere

**Reply by Jdiesel87** [reply]
One could probably write a simple script to hammer it and see what happens

**Reply by GrapeApe230** [reply]
Same here. I remember running into this problem and was left scratching my head why. I prototyped a whole process of identifying audiobooks that had no tags or metadata. Essentially you would feed it "audiobook.mp3" and it would take random samplings of the audio to send it to Mozilla's Project DeepSpeech. It would transcribe the audio samples and submit full-text searches against the google books api to find the book. I burned through my limit pretty quickly.

**Comment by advplyr** [comment]
Thanks to @keaganhilliard, Audible is now available as a provider!

![image](https://user-images.githubusercontent.com/67830747/142780046-d9287443-7bca-445f-9de4-43485927c55a.png)

Audible matches also include the narrator, series and volume number.
